app/upgrades/v1-fork-btc-reorg-k/testnet/upgrade.go

```diff
@@ -0,0 +1,349 @@
+	if err != nil {
+		panic(fmt.Errorf("failed to run the fork handler: %w", err))
+	}
+}
+
+// ForkHandler wraps the logic of the fork to return an error
+func ForkHandler(context sdk.Context, keepers *keepers.AppKeepers) error {
+	ctx := sdk.UnwrapSDKContext(context)
+	l := ctx.Logger()
+
+	// this upgrade should be called when there is a BTC reorg higher than K blocks (btccheckpoint.BtcConfirmationDepth)
+	btcStkK, btcLgtK, finalK := keepers.BTCStakingKeeper, keepers.BTCLightClientKeeper, keepers.FinalityKeeper
+
+	largerBtcReorg := btcStkK.GetLargestBtcReorg(ctx)
+	btcBlockHeightRollbackFrom := largerBtcReorg.RollbackFrom.Height
+
+	l.Debug(
+		"running upgrade due to large BTC reorg",
+		zap.Uint32("btc_block_height_rollback_from", btcBlockHeightRollbackFrom),
+		zap.Uint32("btc_block_height_rollback_to", largerBtcReorg.RollbackTo.Height),
+	)
+
+	btcTip := btcLgtK.GetTipInfo(ctx)
+	paramsByVs := btcStkK.GetAllParamsByVersion(ctx)
+
+	cacheBtcDelByStkTxHashHex := make(map[string]*bstypes.BTCDelegation, 0)
+
+	// collects and deletes the events that are still not processed but
+	// the act that generated the event was rolledbacked:
+	// staking tx or unbonding
+	mapStkTxHashRollback := HandleDeleteVotingPowerDistributionEvts(
+		ctx,
+		&btcStkK,
+		paramsByVs,
+		cacheBtcDelByStkTxHashHex,
+		btcTip.Height,
+		largerBtcReorg,
+		MapUnbondStkTxHashRollback,
+	)
+
+	// unsets the values on BTC delegation and returns the values to update
+	// the voting power table
+	satsToUnbondByFpBtcPk, satsToActivateByFpBtcPk, err := HandleBtcDelegations(
+		ctx,
+		&btcStkK,
+		&finalK,
+		cacheBtcDelByStkTxHashHex,
+		btcTip.Height,
+		paramsByVs,
+		mapStkTxHashRollback,
+		MapUnbondStkTxHashRollback,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to handle BTC delegations: %w", err)
+	}
+
+	// Updates the voting power table accordingly to the BTC delegations rollback actions.
+	HandleVotingPowerDistCache(ctx, &finalK, satsToActivateByFpBtcPk, satsToUnbondByFpBtcPk)
+
+	// deletes the old largest reorg to avoid panic at end blocker again
+	return btcStkK.DeleteLargestBtcReorg(ctx)
+}
+
+func HandleDeleteVotingPowerDistributionEvts(
+	ctx context.Context,
+	btcStkK *bskeeper.Keeper,
+
+	paramsByVs map[uint32]*bstypes.Params,
+	cacheBtcDelByStkTxHashHex map[string]*bstypes.BTCDelegation,
+	btcTipHeight uint32,
+	largerBtcReorg *bstypes.LargestBtcReOrg,
+	mapUnbondStkTxHashRollback map[string]struct{},
+) (mapStkTxHashRollback map[string]struct{}) {
+	mapStkTxHashRollback = make(map[string]struct{}, 0)
+	eventsToDelete := make([]bstypes.EventIndex, 0)
+
+	higherBtcStakingPeriod := MaxBtcStakingTimeBlocks(paramsByVs)
+	btcBlockHeightRollbackFrom := largerBtcReorg.RollbackFrom.Height
+
+	l := sdk.UnwrapSDKContext(ctx).Logger()
+
+	// iterating over all the BTC staking events from the rollback height until latest Tip + the maximum staking period
+	for btcHeight := btcBlockHeightRollbackFrom; btcHeight <= btcTipHeight+higherBtcStakingPeriod; btcHeight++ {
```

Comment:
should this be the `btcBlockHeightRollbackTo` instead of the `btcBlockHeightRollbackFrom`, considering we'll be rollbacking in the range [`btcBlockHeightRollbackTo`, `btcBlockHeightRollbackFrom`]
```suggestion
	for btcHeight := btcBlockHeightRollbackTo; btcHeight <= btcTipHeight+higherBtcStakingPeriod; btcHeight++ {
```

Reply:
Actually, we call ` [btcBlockHeightRollbackFrom, btcBlockHeightRollbackTo]` in the largest BTC reorg [object](https://github.com/babylonlabs-io/babylon/blob/dbcc7c4c72141240f2770754109437c735977401/x/btcstaking/types/btcstaking.pb.go#L1006) as well, now I am questioning myself which one is correct :thinking:

Reply:
you were correct, we tested this yesterday and the `RollbackFrom` is higher than `RollbackTo` :sweat_smile: https://github.com/babylonlabs-io/babylon/blob/0f648f51f4f5b7ad4fed166c3702981e9968b8ff/app/upgrades/v1-fork-btc-reorg-k/testnet/upgrade.go#L69